### PR TITLE
KAB-1057: Add flows to MCP Server

### PR DESCRIPTION
**Pull request by radektomasek**
### Summary

Adds flow (orchestration) management functionality to enable creating and managing data pipelines through the MCP server.

### New Tools Added

- `create_flow` - Create new data flows/pipelines.
- `retrieve_flows` - List all flows or get specific flows by ID.
- `update_flow` - Modify existing flow configurations.
- `get_flow_detail` - Get detailed flow information.

### Files Added

- `keboola_mcp_server/tools/flows.py`

### Key Features

- Schema-compliant with official Keboola flow structure.
- Auto-generates unique IDs for phases and tasks.
- Validates dependencies and prevents circular references.
- Should work with any component type (extractors, transformations, writers) - Would like to do more testing on that front.

### Testing

- Successfully created, retrieved, and updated flows with AWS S3 extractor component

### Video

https://github.com/user-attachments/assets/8ea2aad9-6609-40c2-b66a-10db841c7546

### CODE REVIEW

File: src/keboola_mcp_server/tools/flow.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+    new_raw_configuration = await client.storage_client.create_component_root_configuration(
+        component_id=ORCHESTRATOR_COMPONENT_ID,
+        data={
+            'name': name,
+            'description': description,
+            'configuration': {
+                'parameters': flow_parameters
+            }
+        }
+    )
```
[the remark this answers is not in the document]
Comment (radektomasek) [reply src/keboola_mcp_server/tools/flow.py:104]: 👍

File: src/keboola_mcp_server/tools/flow.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+                        f'branch/{client.storage_client.branch_id}/components/'
+                        f'{ORCHESTRATOR_COMPONENT_ID}/configs/{flow_id}'
+                    )
+                )
+                flows.append(raw_config)
+            except Exception as e:
+                LOG.warning(f'Could not retrieve flow {flow_id}: {e}')
+        return flows
+    else:
+        raw_flows = await client.storage_client.configuration_list(component_id=ORCHESTRATOR_COMPONENT_ID)
```
[the remark this answers is not in the document]
Comment (radektomasek) [reply src/keboola_mcp_server/tools/flow.py:156]: 👍

File: src/keboola_mcp_server/tools/flow.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+@with_session_state()
+async def update_flow(
+    ctx: Context,
+    configuration_id: Annotated[str, Field(description='ID of the flow configuration to update')],
+    name: Annotated[str, Field(description='Updated flow name')],
+    description: Annotated[str, Field(description='Updated flow description')],
+    phases: Annotated[List[Dict[str, Any]], Field(description='Updated list of phase definitions')],
+    tasks: Annotated[List[Dict[str, Any]], Field(description='Updated list of task definitions')],
+    change_description: Annotated[str, Field(description='Description of changes made')],
+) -> Annotated[ComponentRootConfiguration, Field(description='Updated flow configuration')]:
```
[the remark this answers is not in the document]
Comment (radektomasek) [reply src/keboola_mcp_server/tools/flow.py:171]: 👍

File: src/keboola_mcp_server/tools/flow.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+    - Use when you want to see the full details of a specific flow
+
+    EXAMPLES:
+    - user_input: `Show me details of flow 12345`
+        -> set configuration_id to "12345"
+        -> returns the detailed flow configuration
+    """
+    client = KeboolaClient.from_state(ctx.session.state)
+
+    raw_config = await client.storage_client.get(
```
[the remark this answers is not in the document]
Comment (radektomasek) [reply src/keboola_mcp_server/tools/flow.py:240]: 👍

File: src/keboola_mcp_server/tools/flow.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        processed_phases.append(phase_copy)
+
+    return processed_phases
+
+
+def _ensure_task_ids(tasks: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+    """Ensure all tasks have unique IDs and proper structure according to schema"""
+    processed_tasks = []
+    used_ids = set()
+    task_counter = 20001
```
[the remark this answers is not in the document]
Comment (radektomasek) [reply src/keboola_mcp_server/tools/flow.py:287]: This pattern comes from the Kai-Bot implementation that was shared by @davidesner in the KAB-1057: https://github.com/keboola/kai-bot/blob/main/src/keboola/kaibot/backend/flow_backend.py

I just iterated with AI and tried to get some common ground for our implementation (including the incorporation of the Flow model).

The thoughts about this specific code:

```
The Logic:

Avoid conflicts with phase IDs - Phases typically use small numbers (1, 2, 3...).
Clear separation - Tasks get a completely different ID range.
Room for growth - 20,000+ IDs available for tasks.
```

The 20001 starting point is essentially a namespace separation technique - it's a simple, effective way to ensure phase and task IDs never collide while maintaining human-readable sequential numbering.

But in the first place - it was my starting point from the kaibot code. 

I will add some summary to the code itself. @mariankrotil

File: src/keboola_mcp_server/tools/components/model.py [hunk 6]
```diff
@@ -311,3 +313,141 @@ class ComponentWithConfigurations(BaseModel):
...
+            is_disabled=raw_config.get('isDisabled', False),
+            is_deleted=raw_config.get('isDeleted', False),
+            configuration=flow_config,
+            change_description=raw_config.get('changeDescription'),
+            configuration_metadata=raw_config.get('metadata', []),
+            created=raw_config.get('created')
+        )
+
+
+class FlowSummary(BaseModel):
```
Comment (ncapek) [inline src/keboola_mcp_server/tools/components/model.py:412]: We've been using the naming "Reduced" for the listing versions, e.g. "ReducedComponent", "ReducedConfiguration". If possible, can we be consistent with this?

File: src/keboola_mcp_server/tools/components/model.py [hunk 7]
```diff
@@ -311,3 +313,141 @@ class ComponentWithConfigurations(BaseModel):
...
+
+
+class FlowConfiguration(BaseModel):
+    """Represents a complete flow configuration."""
+
+    phases: List[FlowPhase] = Field(description='List of phases in the flow')
+    tasks: List[FlowTask] = Field(description='List of tasks in the flow')
+
+
+class FlowConfigurationResponse(ComponentConfigurationResponseBase):
```
[the remark this answers is not in the document]
Comment (radektomasek) [reply src/keboola_mcp_server/tools/components/model.py:363]: Great point about the response objects. You're absolutely right that mixing core domain objects with API response metadata creates unnecessary complexity.

I initially followed the existing component tools pattern (with the help of AI) for consistency, but I agree this is a good opportunity to improve the architecture.

I am happy to adjust the code and have a cleaner apprach while returning core objects only:

```python
# Instead of FlowConfigurationResponse
def create_flow(...) -> FlowConfiguration
def update_flow(...) -> FlowConfiguration  
def get_flow_detail(...) -> FlowConfiguration
def retrieve_flows(...) -> List[FlowSummary]
```

Let me address it now.

**Comment by radektomasek** [comment]
@ncapek @mariankrotil I've updated the branch and addressed your latest PR comments, including the schema adjustment that @davidesner [mentioned in the ticket](https://keboola.atlassian.net/browse/KAB-1057?focusedCommentId=163978).

## Changes Made:
- ✅ **Fixed configuration structure** - Removed extra `parameters` wrapper (flows store phases/tasks directly under `configuration`).
- ✅ **Added comprehensive schema documentation** - Embedded the simplified flow schema in tool descriptions so LLMs have clear guidance.
- ✅ **Updated model parsing** - Fixed parsing to read from correct API response location.
- ✅ **Clean architecture maintained** - Tools return core objects only (`FlowConfiguration`, `List[ReducedFlow]`).

## Testing:
- **Local testing successful** - Flows create, retrieve, and update correctly using Claude as the MCP client.
- **Unit tests added** - Comprehensive test coverage including edge cases, validation logic, and integration scenarios.
- **All major functionality verified** - ID generation, circular dependency detection, and API integration working as expected.

The implementation now works "like a charm" as David mentioned - the LLM receives proper schema guidance and creates the correct API structure without guessing.

Ready for handover if you're satisfied with the changes! 🚀

cc @jordanrburger

File: src/keboola_mcp_server/tools/flow.py [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+import logging
+from typing import Annotated, Any, Dict, List, Optional
+
+from fastmcp import Context, FastMCP
+from pydantic import Field
+
+from keboola_mcp_server.client import KeboolaClient
+from keboola_mcp_server.errors import tool_errors
+from keboola_mcp_server.mcp import with_session_state
+from keboola_mcp_server.tools.components.model import (
```
Comment (mariankrotil) [inline src/keboola_mcp_server/tools/flow.py:12]: As the files are getting large, We could create a new tools submodule: `tools/flows` where we could have `flows/model.py` - flow models, `flows/utils.py` - validation methods and `flows/tools.py` - tool implementations like it is done with components?

File: src/keboola_mcp_server/tools/flow.py [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+
+@tool_errors()
+@with_session_state()
+async def create_flow(
+    ctx: Context,
+    name: Annotated[str, Field(description='A short, descriptive name for the flow')],
+    description: Annotated[str, Field(description='Detailed description of the flow purpose')],
+    phases: Annotated[List[Dict[str, Any]], Field(description=f"""List of phase definitions.
+
+{FLOW_SCHEMA}
```
Comment (mariankrotil) [inline src/keboola_mcp_server/tools/flow.py:88]: We should not add schemas to the descriptions as when the agent wants to list the tools, this tool description gets extremely large and can shadow others. @ncapek suggested tool for getting flow schmeas

File: src/keboola_mcp_server/tools/flow.py [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+    phases: Annotated[List[Dict[str, Any]], Field(
+        description=f"""Updated list of phase definitions.
+            {FLOW_SCHEMA}
+            """
+        )],
+        tasks: Annotated[List[Dict[str, Any]], Field(
+            description=f"""Updated list of task definitions.
+                {FLOW_SCHEMA}"""
+        )],
+        change_description: Annotated[str, Field(description='Description of changes made')],
```
Comment (mariankrotil) [inline src/keboola_mcp_server/tools/flow.py:144]: align the indentation, please

File: src/keboola_mcp_server/client.py [hunk 11]
```diff
@@ -524,6 +526,85 @@ async def update_component_row_configuration(
...
+
+    async def list_flow_configurations(self) -> list[JsonDict]:
+        """
+        Lists all flow (orchestrator) configurations in the project.
+
+        :return: List of flow configurations
+        """
+        return await self.configuration_list(component_id=ORCHESTRATOR_COMPONENT_ID)
+
+    async def get_flow_configuration(self, config_id: str) -> JsonDict:
```
Comment (mariankrotil) [inline src/keboola_mcp_server/client.py:596]: Do we need to implement those client methods for flows (the same for `update_flow_configuration` and others), within our flow tools we can just use: 
```
await client.configuration_detail(
            component_id=ORCHESTRATOR_COMPONENT_ID,
            configuration_id=config_id
        )
  ```

File: src/keboola_mcp_server/tools/flow.py (hunk 8 above)
Comment (vita-stejskal) [reply src/keboola_mcp_server/tools/flow.py:12]: Let's leave this refactoring to another PR.

File: src/keboola_mcp_server/tools/flow.py (hunk 9 above)
Comment (vita-stejskal) [reply src/keboola_mcp_server/tools/flow.py:88]: This is addressed by #124 .

File: src/keboola_mcp_server/client.py (hunk 11 above)
Comment (vita-stejskal) [reply src/keboola_mcp_server/client.py:596]: There are certainly pros and cons of this approach:

* pros -- more compact code in the flow tools; easier mocking in the tests
* cons -- clients cluttered with many function

Anyway, let's leave it as it is.

File: src/keboola_mcp_server/tools/flow.py [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+
+
+@tool_errors()
+@with_session_state()
+async def retrieve_flows(
+    ctx: Context,
+    flow_ids: Annotated[
+        Sequence[str],
+        Field(default_factory=tuple, description='The configuration IDs of the flows to retrieve.')
+    ] = tuple(),
```
[the remark this answers is not in the document]
Comment (vita-stejskal) [reply src/keboola_mcp_server/tools/flow.py:179]: The python default is needed when calling the function `retrieve_flows(...)` directly from python, e.g. in tests [here](https://github.com/keboola/mcp-server/blob/KAB-1057-flow-get-update-tool/tests/tools/test_flows.py#L423).

I am not sure if `default_factory` is really needed here, but I think that it affects the tool's "input JSON schema" and perhaps also the way how the tool's function is called via the MCP protocol. I admit that my understanding is limited in this area.

**Review by vita-stejskal — APPROVED** [review]
